### PR TITLE
fix #636: mitigate long header attack

### LINKED ISSUE #636: Make use of MaxHeaderBytes in REST API servers.
#### Pre-check

* [x] This is not a IPFS Cluster website content issue ([file those here](https://github.com/ipfs/ipfs-cluster-website/issues))
* [x] I read the [troubleshooting section](https://cluster.ipfs.io/documentation/troubleshooting/) of the website and it did not help
* [x] I searched for similar issues in the repo without luck
* [x] All my peers are running the same cluster version
* [x] All my peers are configured using the same cluster `secret`

#### Basic information

* [x] Version information (mark as appropiate):
  * [x] Master
  * [ ] Release candidate for next version
  * [ ] Latest stable version
  * [ ] An older version I should not be using
* [x] Type (mark as appropiate):
  * [ ] Bug
  * [x] Feature request
  * [x] Enhancement

#### Description

A malicious client can DoS the API by posting very long headers/request line. We should limit to something acceptable like 4096 by default (but configurable like the other server options). Worth for proxy server too.

### PR DESCRIPTION
**Pull request by alekswn**
License: MIT
Signed-off-by: Alexey Novikov <alexey@novikov.io>

### CODE REVIEW

File: api/ipfsproxy/config_test.go [hunk 1]
```diff
@@ -14,13 +14,22 @@ var cfgJSON = []byte(`
       "read_timeout": "10m0s",
       "read_header_timeout": "5s",
       "write_timeout": "10m0s",
-      "idle_timeout": "1m0s",
+	  "idle_timeout": "1m0s",
```
Comment (lanzafame) [inline api/ipfsproxy/config_test.go]:
```suggestion
	"idle_timeout": "1m0s",
```

File: api/ipfsproxy/config_test.go [hunk 2]
```diff
@@ -14,13 +14,22 @@ var cfgJSON = []byte(`
...
+	  "max_header_bytes": 16384,
```
Comment (lanzafame) [inline api/ipfsproxy/config_test.go]:
```suggestion
	"max_header_bytes": 16384,
```

File: api/rest/config_test.go [hunk 3]
```diff
@@ -20,7 +20,8 @@ var cfgJSON = []byte(`
       "read_timeout": "30s",
       "read_header_timeout": "5s",
       "write_timeout": "1m0s",
-      "idle_timeout": "2m0s",
+	  "idle_timeout": "2m0s",
```
Comment (lanzafame) [inline api/rest/config_test.go]:
```suggestion
	"idle_timeout": "2m0s",
```

File: api/rest/config_test.go [hunk 4]
```diff
@@ -20,7 +20,8 @@ var cfgJSON = []byte(`
...
+	  "max_header_bytes": 16384,
```
Comment (lanzafame) [inline api/rest/config_test.go]:
```suggestion
	"max_header_bytes": 16384,
```

File: api/rest/restapi_test.go [hunk 5]
```diff
@@ -1073,21 +1074,27 @@ type httpTestcase struct {
...
 }
 
-func assertHTTPStatusIsNotUnauthoriazed(resp *http.Response) error {
-	if assertHTTPStatusIsUnauthoriazed(resp) == nil {
-		return fmt.Errorf("unexpected HTTP status code: %d", http.StatusUnauthorized)
+func assertHTTPStatusIsTooLarge(resp *http.Response) error {
+	return httpStatusCodeChecker(resp, http.StatusRequestHeaderFieldsTooLarge)
+}
+
+func makeInvertedHTTPStatusAssert(checker responseChecker) responseChecker {
```
Comment (lanzafame) [inline api/rest/restapi_test.go]:
```suggestion
func makeHTTPStatusNegatedAssert(checker responseChecker) responseChecker {
```
Just to be clear it is the boolean aspect of the assert that is being inverted and not the status code.

**Comment by hsanjuan** [comment]
@lanzafame,  if you approve you can go ahead and merge

**Pull request merged**